Dev nearby

Prashant-Android opened the pull request:
This PR adds better handling for IP addresses and refines the MMCP Ping and Pong message flow. It also includes improvements in connection management.

mikedawson committed:
Add NetworkInterface interfaces

Prashant-Android committed:
nearby

Prashant-Android committed:
nearby testing

Prashant-Android committed:
Almost complete only remaining error is "BufferUnderflowException"

Prashant-Android committed:
nearby discovery and connection

Prashant-Android committed:
nearby discovery and connection, sharing ip

Prashant-Android committed:
nearby discovery and connection, sharing ip

Prashant-Android committed:
Improve IP handling and MMCP message flow

mikedawson commented on lib-nearby/src/main/java/com/meshrabiya/lib_nearby/nearby/NearbyVirtualNetwork.kt:
```diff
@@ -0,0 +1,351 @@
+        }
+    }
+
+    private fun checkAndInitiateConnections(endpointMap: Map<String, EndpointInfo>) {
```
There is no reason for this to be a separate function. It is a private function, not very long, and only ever called by line 220 (observeEndpointStatusFlow). I think this makes the code harder to read/follow.

mikedawson commented on lib-nearby/src/main/java/com/meshrabiya/lib_nearby/nearby/NearbyVirtualNetwork.kt:
```diff
@@ -0,0 +1,351 @@
+    }
+
+    private fun observeEndpointStatusFlow() {
+        CoroutineScope(Dispatchers.Default).launch {
```
You need to create a scope (as a member val) which should then be canceled by calling close.

mikedawson commented on lib-nearby/src/main/java/com/meshrabiya/lib_nearby/nearby/NearbyVirtualNetwork.kt:
```diff
@@ -0,0 +1,351 @@
+
+    private fun log(message: String, exception: Exception? = null) {
+        val prefix = "[NearbyVirtualNetwork:$name] "
+        if (exception != null) {
```
This function must take an argument for the log level. You must ensure that log calls use an appropriate log level e.g. more important less frequent events like connecting and disconnecting using a high priority such as info, very frequent events (like receiving a packet) using debug/verbose levels.

You need to set appropriate verbosity now as you write it such that it can be filtered as needed when the log output is being used for debugging.

mikedawson commented on lib-nearby/src/main/java/com/meshrabiya/lib_nearby/nearby/NearbyVirtualNetwork.kt:
```diff
@@ -0,0 +1,351 @@
+    private val endpointDiscoveryCallback = object : EndpointDiscoveryCallback() {
+        override fun onEndpointFound(endpointId: String, info: DiscoveredEndpointInfo) {
+            log("Endpoint found: $endpointId")
+            addEndpoint(endpointId)
```
No reason to have addEndpoint as its own function

mikedawson commented on lib-nearby/src/main/java/com/meshrabiya/lib_nearby/nearby/NearbyVirtualNetwork.kt:
```diff
@@ -0,0 +1,351 @@
+        override fun onEndpointFound(endpointId: String, info: DiscoveredEndpointInfo) {
+            log("Endpoint found: $endpointId")
+            addEndpoint(endpointId)
+            checkAndInitiateConnection(endpointId)
```
Connections should __only__ be initiated by observing the status. We could discover more nodes than we want to connect to,  this function will _wrongly_ attempt to connect to every single node discovered.

mikedawson commented on lib-nearby/src/main/java/com/meshrabiya/lib_nearby/nearby/NearbyVirtualNetwork.kt:
```diff
@@ -0,0 +1,351 @@
+
+    private fun sendMmcpPingPacket(endpointId: String) {
+        val mmcpPing = MmcpPing(Random.nextInt())
+        val pingBytes = mmcpPing.toBytes()
```
Please use the MmcpMessage.toVirtualPacket function - that's already there and tested.

mikedawson commented on lib-nearby/src/main/java/com/meshrabiya/lib_nearby/nearby/NearbyVirtualNetwork.kt:
```diff
@@ -0,0 +1,351 @@
+        val totalSize = VirtualPacketHeader.HEADER_SIZE + pingBytes.size
+        val data = ByteArray(totalSize)
+        header.toBytes(data, 0)
+        System.arraycopy(pingBytes, 0, data, VirtualPacketHeader.HEADER_SIZE, pingBytes.size)
```
No need to make a copy of the packet. That does have some (small) overhead, but as many packets are being exchanged on a regular basis, better to keep that overhead to a minimum.

mikedawson commented on lib-nearby/src/main/java/com/meshrabiya/lib_nearby/nearby/NearbyVirtualNetwork.kt:
```diff
@@ -0,0 +1,351 @@
+        val mmcpPong = MmcpPong(Random.nextInt(), replyToMessageId)
+        val pongBytes = mmcpPong.toBytes()
+
+        val header = VirtualPacketHeader(
```
As above - change to using MmcpMessage's .toVirtualPacket, don't use system arraycopy

mikedawson commented on lib-nearby/src/main/java/com/meshrabiya/lib_nearby/nearby/NearbyVirtualNetwork.kt:
```diff
@@ -0,0 +1,351 @@
+            .addOnFailureListener { e -> log("Failed to send MMCP Pong to $endpointId", e) }
+    }
+
+    private fun Int.toByteArray(): ByteArray {
```
Please do not duplicate functions we already have that are already there, tested, and consistented. Look at the existing code - e.g. VirtualNode and use the existing functions.

mikedawson commented on lib-nearby/src/main/java/com/meshrabiya/lib_nearby/nearby/NearbyVirtualNetwork.kt:
```diff
@@ -0,0 +1,351 @@
+
+    data class EndpointInfo(
+        val endpointId: String,
+        var status: EndpointStatus = EndpointStatus.DISCONNECTED,
```
These need to be val. The idea of flows is that everything is an immutable val, if it is changed, then we should use .copy. Otherwise something that is observing the flow won't see the change.

Prashant-Android committed:
Improve IP handling and MMCP message flow

mikedawson commented on lib-nearby/src/main/java/com/meshrabiya/lib_nearby/nearby/NearbyVirtualNetwork.kt:
```diff
@@ -0,0 +1,286 @@
+        observeEndpointStatusFlow()
+    }
+
+    fun start() {
```
Why not start advertising and discovery directly in init? That means that this interface won't work until calling start - which makes it more complicated for anyone using this library.

mikedawson commented on lib-nearby/src/main/java/com/meshrabiya/lib_nearby/nearby/NearbyVirtualNetwork.kt:
```diff
@@ -0,0 +1,286 @@
+        startDiscovery()
+    }
+
+    fun stop() {
```
This should be close. It should also use an atomic boolean to make sure that even if close is called multiple times, this function should only be run once:

e.g. 

val isClosed = AtomicBoolean(false)

if(!isClosed.getAndSet(true)) {
   //stop advertising discovery scope etc 
}

mikedawson replied:
Close functions should be stable even if they are called multiple times - hence you need the atomic boolean.

Prashant-Android committed:
Improve IP handling and MMCP message flow

Prashant-Android committed:
fixed

Prashant-Android committed:
Chat functionality working but text are not properly showing.

Prashant-Android committed:
Chat functionality working.

Prashant-Android committed:
Chat functionality working.

Prashant-Android committed:
Nearby Stream added

mikedawson commented on lib-nearby/src/main/java/com/meshrabiya/lib_nearby/nearby/NearbyStreamHeader.kt:
```diff
@@ -0,0 +1,28 @@
+
+import java.nio.ByteBuffer
+
+data class NearbyStreamHeader(
```
This should extend MmcpMessage - we also need the info included in it / virtual packet e.g. from address, to address, etc.

mikedawson commented on lib-nearby/src/main/java/com/meshrabiya/lib_nearby/nearby/NearbyStreamHeader.kt:
```diff
@@ -0,0 +1,28 @@
+    fun toBytes(): ByteArray {
+        return ByteBuffer.allocate(12).apply {
+            putInt(streamId)
+            put(if (isReply) 1.toByte() else 0.toByte())
```
no need for the if statement - Boolean.toByte() will already do the same

mikedawson commented on lib-nearby/src/main/java/com/meshrabiya/lib_nearby/nearby/NearbyVirtualNetwork.kt:
```diff
@@ -0,0 +1,331 @@
+    private fun handleStreamPayload(endpointId: String, payload: Payload) {
+        checkClosed()
+        payload.asStream()?.asInputStream()?.use { inputStream ->
+            val headerBytes = ByteArray(12)
```
By convention never use any literal hardcoded like this other than 0, 1, true, and false. This should be defined as a single constant. If something was added this will get bigger. There should be one and only one place that should be changed.

Important principle - DRY - Dont Repeat Yourself / use a single point of truth.

Prashant-Android committed:
Sending UDP packets as a byte Payload.

Prashant-Android committed:
Completed

Prashant-Android committed:
Module lib-meshrabiya-vpn Added

Prashant-Android committed:
Module lib-meshrabiya-vpn Added

Prashant-Android committed:
App traffic is routed and Internet works as normal fixed.

Prashant-Android committed:
App can filter/examine each packet and determine/log

mikedawson commented on lib-meshrabiya-vpn/src/main/java/com/meshrabiya/lib_vpn/MeshrabiyaVpnService.kt:
```diff
@@ -0,0 +1,79 @@
+    private fun establishVpn() {
+        try {
+            builder.apply {
+                addAddress("10.0.0.2", 24)
```
Does not adhere to coding style guideline: hardcoded literals

mikedawson commented on lib-meshrabiya-vpn/src/main/java/com/meshrabiya/lib_vpn/MeshrabiyaVpnService.kt:
```diff
@@ -0,0 +1,240 @@
+
+            // Establish the VPN interface
+            vpnInterface = builder.establish()
+            vpnInterface?.let {
```
This is silently failing - as per https://github.com/UstadMobile/UstadMobile/blob/primary/CODING-STYLE.md "Do not use null checks that fail silently and would lead to code that doesn't behave as expect with no logging output or exception" - if vpninterface was null - this would fail silently.

Prashant-Android committed:
App can filter/examine each packet and determine/log

Prashant-Android committed:
- Migrated dependency management to `libs.versions.toml`
- Updated dependencies to match versions used in Ustad.

Prashant-Android committed:
- Migrated dependency management to `libs.versions.toml`
- Updated dependencies to match versions used in Ustad.

Prashant-Android committed:
- Replaced hardcoded literals with constants and utilized InetAddress for IP addresses.
- Replaced direct `Thread` usage with `ExecutorService` for better thread management.
- Ensured all IP address handling uses `InetAddress` instead of raw `String`.
- Refactored header parsing functions to handle exceptions and follow coding conventions.

mikedawson commented on lib-meshrabiya-vpn/src/main/java/com/meshrabiya/lib_vpn/MeshrabiyaVpnService.kt:
```diff
@@ -0,0 +1,306 @@
+        private const val IPPROTO_ICMPV6 = 58 // ICMPv6 protocol number
+
+        // Constants for byte offsets and field lengths
+        private const val IPV4_PROTOCOL_OFFSET = 9
```
Needs a comment on each constant explaining why that is the value or link to reference.

mikedawson commented on lib-meshrabiya-vpn/src/main/java/com/meshrabiya/lib_vpn/MeshrabiyaVpnService.kt:
```diff
@@ -0,0 +1,306 @@
+     * @param length the length of the packet data.
+     */
+    private fun processPacket(buffer: ByteBuffer, length: Int) {
+        buffer.flip()
```
Why flip it? Needs an explanation / reference as to why that is being done.

mikedawson commented on lib-meshrabiya-vpn/src/main/java/com/meshrabiya/lib_vpn/MeshrabiyaVpnService.kt:
```diff
@@ -0,0 +1,306 @@
+    private fun parseIPv6Header(buffer: ByteBuffer): Pair<Int, String> {
+        try {
+            val protocol = buffer.get(IPV6_PROTOCOL_OFFSET).toInt()
+            val isTcp = protocol == IPPROTO_TCP
```
Protocol: would be better to use an enum for this because there are only three valid values.

mikedawson commented on lib-meshrabiya-vpn/src/main/java/com/meshrabiya/lib_vpn/MeshrabiyaVpnService.kt:
```diff
@@ -0,0 +1,306 @@
+     * @param buffer the buffer containing the IPv6 packet data.
+     * @return a pair of the destination port and the protocol name.
+     */
+    private fun parseIPv6Header(buffer: ByteBuffer): Pair<Int, String> {
```
parseIpHeader functions should return a data class (e.g. IpHeader) - that should contain destination address (InetAddress), destination port, origin address (InetAddress), origin port, and protocol.

mikedawson replied:
Note we need to know the origin port because there could be multiple streams going to the same destination. For Meshrabiya to match things up correctly, we will need to know the origin.

Prashant-Android committed:
Added explanations for constants, clarified buffer flip logic, updated protocol handling with enum, and modified parseIpHeader to return a data class with origin details

Prashant-Android committed:
Optimize code structure and improve adherence to coding standards for ChatServer.

mikedawson committed:
Add TODOs

Prashant-Android committed:
Implemented VirtualNetworkInterface.

Prashant-Android committed:
Made necessary changes in NearbyVirtualNetwork. Updated the Chat server according to the changes in NearbyVirtualNetwork. Added a method in OriginatingMessageManager to get the list of virtual IP addresses.

Prashant-Android committed:
Optimized code in ChatServer class, ViewModel, and NearbyTestScreen.

mikedawson committed:
Add todo on virtual message manager

Prashant-Android committed:
- Update VirtualNode to use multiple virtual interfaces
- Modify OriginatingMessageManager to handle multiple addresses

Prashant-Android committed:
- Update VirtualNode to use multiple virtual interfaces
- Modify OriginatingMessageManager to handle multiple addresses

Prashant-Android committed:
- Update VirtualNode to use multiple virtual interfaces
- Modify OriginatingMessageManager to handle multiple addresses

Prashant-Android committed:
- Update VirtualNode to use multiple virtual interfaces
- Modify OriginatingMessageManager to handle multiple addresses

Prashant-Android committed:
- Update VirtualNode to use multiple virtual interfaces
- Modify OriginatingMessageManager to handle multiple addresses

Prashant-Android committed:
- Update VirtualNode to use multiple virtual interfaces
- Modify OriginatingMessageManager to handle multiple addresses

Prashant-Android committed:
-added sendUdpPacket function in NearbyVirtualNetwork

Prashant-Android committed:
-added sendUdpPacket function in NearbyVirtualNetwork

Prashant-Android committed:
- Removed unused code

Prashant-Android committed:
- Removed ViewModel from DI setup.
- Added virtualNode.addVirtualNetworkInterface to NearbyTestViewModel.
- Fixed issues in NearbyVirtualNetwork class and added comments for function clarity.
- Added addVirtualNetworkInterface function in VirtualNode.kt.

mikedawson commented on test-app/src/main/java/com/ustadmobile/meshrabiya/testapp/viewmodel/NearbyTestViewModel.kt:
```diff
@@ -0,0 +1,273 @@
+
+    private fun initializeNearbyNetwork() {
+        try {
+            nearbyNetwork = NearbyVirtualNetwork(
```
This **should not** be done here. Initializing the virtual network every time a viewmodel starts is obviously wrong.

mikedawson committed:
Remove VirtualNetwork initialization from ViewModel - move to App and UseCase

mikedawson committed:
Restore start function on NearbyVirtualNetwork
